In [271]:
import pandas as pd
import numpy as np
import random 
data_path = "./cb_db_csv/"

In [2]:
pd.__version__

'0.23.4'

# cb_objects

In [11]:
df_object = pd.read_csv(data_path + 'cb_objects.csv', encoding = "ISO-8859-1")
df_object = df_object.replace(r'\N', np.nan)
df_object.founded_at = pd.to_datetime(df_object.founded_at)
df_object=df_object[pd.notnull(df_object['founded_at'])]
entity_type = ['Company']
df_object=df_object[df_object['entity_type'].isin(entity_type)]
df_object.category_code.unique()

17118087


array(['web', 'network_hosting', 'games_video', 'advertising',
       'cleantech', 'enterprise', 'other', 'consulting', 'mobile',
       'health', 'software', 'analytics', 'finance', 'education',
       'medical', 'manufacturing', 'ecommerce', 'public_relations',
       'hardware', nan, 'search', 'news', 'government', 'security',
       'biotech', 'photo_video', 'travel', 'semiconductor', 'social',
       'legal', 'transportation', 'hospitality', 'sports', 'nonprofit',
       'fashion', 'messaging', 'music', 'automotive', 'design',
       'real_estate', 'local', 'nanotech', 'pets'], dtype=object)

## Helper Method

## IPOs

In [389]:
df_ipo = pd.read_csv(data_path + 'cb_ipos.csv', encoding = "ISO-8859-1")
df_ipo = df_ipo.drop(['source_url','updated_at'],axis=1)
df_ipo = df_ipo.replace(r'\N', np.nan)
## drop object_id = null
## drop companies without id record
df_ipo = df_ipo[df_ipo.object_id.notnull()]

df_ipo = pd.merge(df_ipo,df_object[['id','founded_at']],how='left',left_on=['object_id'],right_on=['id'])


## make timestamp 
df_ipo.public_at = pd.to_datetime(df_ipo.public_at)
df_ipo.public_at = pd.to_datetime(df_ipo.created_at)
df_ipo['time_since_founded'] = pd.Timedelta('nan')
null_columns = df_ipo.columns[df_ipo.isnull().any()]
print(df_ipo[null_columns].isnull().sum())
df_ipo.shape

valuation_amount           1146
valuation_currency_code       2
raised_amount              1115
raised_currency_code        558
id_y                        547
founded_at                  547
time_since_founded         1254
dtype: int64


(1254, 13)

- Before: 1254
- Drop non-existing company id : 707


In [390]:
def getIPODuration(row):
    if pd.isnull(row['public_at']):
        row['public_at'] = row['created_at']
    if pd.notnull(row['founded_at']) and pd.notnull(row['public_at']):
        duration = row['public_at'] - row.founded_at
        row['time_since_founded'] = pd.to_timedelta([duration]).astype('timedelta64[D]').values
    return row
df_ipo = df_ipo.apply(getIPODuration, axis=1)
df_ipo = df_ipo[df_ipo['time_since_founded'].notnull()]
df_ipo = df_ipo.drop(['id_x','id_y','ipo_id','valuation_amount','valuation_currency_code','raised_amount','raised_currency_code',
                      'public_at','stock_symbol','created_at','founded_at'],axis=1)
df_ipo.sort_values(by=['time_since_founded'],inplace=True)
df_ipo = df_ipo[df_ipo['time_since_founded'].gt(-0.1)]

In [391]:
df_ipo

object_id time_since_founded
27      c:4495                [1]
908   c:144059               [67]
11       c:475               [83]
198    c:10614              [121]
806   c:164413              [324]
691   c:145411              [340]
1224  c:281988              [694]
1253  c:235945              [711]
716    c:48759              [754]
611    c:22130              [896]
973   c:245010              [947]
1017  c:258641              [968]
1094  c:265792              [992]
309     c:1368             [1018]
229    c:37881             [1081]
681    c:11391             [1107]
709    c:24982             [1168]
15       c:108             [1197]
658    c:40482             [1253]
957   c:241173             [1301]
963    c:46007             [1306]
1034  c:258957             [1334]
867    c:37431             [1409]
1214   c:37940             [1414]
486     c:1448             [1448]
65     c:15609             [1476]
308    c:43666             [1531]
936    c:38270             [1598]
344     c:1249             [1636]
699     c:1812             [1695]
...        ...                ...
124      c:355            [25776]
508    c:64802            [26290]
259    c:39278            [27400]
727    c:34406            [27493]
431    c:58207            [27672]
216    c:37380            [27738]
1250   c:39330            [28470]
584     c:4304            [29679]
646    c:30927            [29811]
1229  c:223939            [29915]
753     c:1915            [30787]
2        c:342            [30797]
414    c:20016            [30933]
1163  c:274050            [31708]
428     c:7576            [32233]
166    c:34181            [32433]
537      c:541            [32533]
14      c:3658            [33323]
140    c:32510            [33504]
1246  c:258527            [33947]
550    c:30981            [34009]
1225   c:27295            [34662]
987    c:57929            [35654]
1022   c:26486            [36033]
746    c:59312            [36993]
710     c:7060            [38791]
367     c:4199            [39642]
1109   c:11318            [40280]
935    c:70725            [40374]
1027   c:20211            [40781]

[705 rows x 2 columns]

In [392]:
df_ipo.to_pickle('./dataframe_cleaned/df_ipo') 

## Acquisition
### Acquistion date: created date 28 companies

In [ ]:
df_acq = pd.read_csv(data_path + 'cb_acquisitions.csv', encoding = "ISO-8859-1")
df_acq = df_acq.drop(['source_url','updated_at'],axis=1)

df_acq = df_acq.replace(r'\N', np.nan)
df_acq = df_acq[df_acq['acquired_object_id'].notnull()]

## Add acquistion data
df_acq.loc[df_acq['acquired_at'].isnull(),'acquired_at'] = df_acq.loc[df_acq['acquired_at'].isnull(),'created_at']
null_columns=df_acq.columns[df_acq.isnull().any()]
print(df_acq[null_columns].isnull().sum())


df_acq = pd.concat([df_acq, pd.get_dummies(df_acq['price_currency_code'])], axis=1, sort=False)
df_acq = pd.concat([df_acq, pd.get_dummies(df_acq['term_code'])], axis=1, sort=False)
df_acq = pd.merge(df_acq,df_object[['id','category_code']],how='left',left_on=['acquired_object_id'],right_on=['id'])

df_acq

In [20]:
## Acquiring
df_acq.acquired_at = pd.to_datetime(df_acq.acquired_at)
df_acq['acquiring_time_since_founded'] = pd.Timedelta('nan')
def getAcquiringDuration(row):
    df_row = df_object.loc[df_object['id'] == row['acquiring_object_id']]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['acquired_at']):
            duration = row['acquired_at'] - df_row['founded_at']
            row['acquiring_time_since_founded'] = duration.astype('timedelta64[D]').values
    return row
df_acq = df_acq.apply(getAcquiringDuration, axis=1)


In [21]:
## Acquired
df_acq['acquired_time_since_founded'] = pd.Timedelta('nan')
def getAcquiredDuration(row):
    df_row = df_object.loc[df_object['id'] == row['acquired_object_id']]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['acquired_at']):
            duration = row['acquired_at'] - df_row['founded_at']
            row['acquired_time_since_founded'] = duration.astype('timedelta64[D]').values
    return row
df_acq = df_acq.apply(getAcquiredDuration, axis=1)


In [22]:
df_acq.to_pickle('./dataframe_cleaned/df_acq') 
df_acq

id_x  acquisition_id acquiring_object_id acquired_object_id  \
0        1               1                c:11               c:10   
1        2               7                c:59               c:72   
2        3               8                c:24              c:132   
3        4               9                c:59              c:155   
4        5              10               c:212              c:215   
5        6              11               c:227              c:228   
6        7              12                 c:5              c:230   
7        8              14               c:318              c:314   
8        9              15                c:20              c:333   
9       10              16               c:342              c:341   
10      11              17                c:24              c:347   
11      12              18               c:349              c:348   
12      13              19               c:355              c:354   
13      14              20               c:377              c:100   
14      15              22               c:351              c:366   
15      16              23               c:307              c:306   
16      17              24               c:212              c:394   
17      18              26               c:499              c:495   
18      19              27                c:29              c:528   
19      20              28                c:29              c:530   
20      21              29               c:541               c:52   
21      22              30               c:541              c:542   
22      23              31               c:568              c:201   
23      24              32               c:570              c:569   
24      25              33                c:20              c:571   
25      26              34               c:317              c:572   
26      27              35               c:573              c:574   
27      28              36               c:583              c:581   
28      29              37               c:584              c:585   
29      30              38               c:590              c:588   
...    ...             ...                 ...                ...   
9531  9533           10496            c:230880           c:285385   
9532  9534           10497             c:50953            c:26342   
9533  9535           10498            c:285396           c:285399   
9534  9536           10499               c:129           c:142814   
9535  9537           10500             c:42365            c:79582   
9536  9538           10501            c:258527           c:285601   
9537  9539           10502            c:258527           c:285603   
9538  9540           10503              c:1434            c:60436   
9539  9541           10505            c:233530           c:285623   
9540  9542           10506            c:285627           c:285630   
9541  9543           10507                c:29           c:169154   
9542  9544           10511            c:205470            c:45837   
9543  9545           10512            c:195465           c:195464   
9544  9546           10513             c:43761            c:43761   
9545  9547           10514            c:285827           c:146509   
9546  9548           10515             c:22784           c:237136   
9547  9549           10516              c:1600            c:71968   
9548  9550           10517            c:285921            c:71996   
9549  9551           10518             c:16591           c:209849   
9550  9552           10519             c:10120           c:285956   
9551  9553           10520               c:672           c:285961   
9552  9554           10521             c:16259           c:283148   
9553  9555           10522            c:285964            c:13710   
9554  9556           10523            c:285968           c:285969   
9555  9557           10524            c:285984           c:285991   
9556  9558           10525            c:267859           c:285995   
9557

In [149]:
## Acquiring

df_acquiring = df_acq.set_index(['acquiring_object_id','acquired_object_id']).sort_index()
df_acquiring.sort_values(by=['acquiring_time_since_founded'],inplace=True)
df_acquiring.drop(['id_x','acquisition_id','acquired_at','acquired_time_since_founded','id_y'],axis=1,inplace=True)
df_acquiring = df_acquiring[df_acquiring['acquiring_time_since_founded'].notnull()]
df_acquiring = df_acquiring[df_acquiring['acquiring_time_since_founded'].gt(0)]

## category average price

In [150]:
category = df_acquiring[df_acquiring['category_code'].notnull()]['category_code'].unique()
avg_price = []
for ctg in category:
    row = df_acquiring.loc[df_acquiring["category_code"]==ctg,['price_amount']]
    price = row[pd.notnull(row.price_amount)].astype('float128').mean()
    avg_price.append(price)
price_dic = dict(zip(category,avg_price))

In [152]:
def getPrice(row):
    if row[['price_amount']].isnull().item() and row[['category_code']].notnull().item():
        row[['price_amount']] = price_dic[row['category_code']]
    return row
df_acquiring = df_acquiring.apply(getPrice,axis=1)
df_acquiring = df_acquiring.dropna(subset=['category_code','price_amount'],how='all')

In [153]:
df_acquiring = df_acquiring[df_acquiring['price_amount'].notnull()]

df_acquiring = df_acquiring.drop(['price_currency_code','term_code','created_at','category_code'],axis=1)

In [166]:
df_acquiring


price_amount  AUD  CAD  EUR  GBP  JPY  \
acquiring_object_id acquired_object_id                                         
c:260783            c:80142             2.40649e+08    0    0    0    0    0   
c:307               c:306               2.39394e+08    0    0    0    0    0   
c:60326             c:146208                1000000    0    0    0    0    0   
c:15461             c:15460                   70000    0    0    0    0    0   
c:192899            c:193355                 330000    0    0    0    0    0   
c:138299            c:141275                  15000    0    0    0    0    0   
c:80259             c:4716              2.40649e+08    0    0    0    0    0   
c:166448            c:40101             2.12816e+08    0    0    0    0    0   
c:82144             c:60638              3.4151e+08    0    0    0    0    0   
c:5457              c:5456                     5000    0    0    0    0    0   
c:64703             c:59456             3.83585e+08    0    0    0    0    0   
c:195859            c:195858                 100000    0    0    0    0    0   
c:221698            c:190514                   5000    0    0    0    0    0   
c:171432            c:44839                    3000    0    0    0    0    0   
c:47470             c:23476              4.0615e+08    0    0    0    0    0   
c:23010             c:110               2.94595e+08    0    0    0    0    0   
c:279935            c:169569            2.40649e+08    0    0    0    0    0   
c:85143             c:47064             2.40649e+08    0    0    0    0    0   
c:207479            c:150485            2.94595e+08    0    0    0    0    0   
c:15180             c:5783              3.02894e+08    0    0    0    0    0   
c:190559            c:38003             2.94595e+08    0    0    0    0    0   
c:84204             c:43263             2.94595e+08    0    0    0    0    0   
c:45891             c:45891             2.94595e+08    0    0    0    0    0   
c:175916            c:25200             2.94595e+08    0    0    0    0    0   
c:64703             c:71792             5.86052e+08    0    0    0    0    0   
c:193054            c:42069             2.40649e+08    0    0    0    0    0   
c:45772             c:45773             2.94595e+08    0    0    0    0    0   
c:257213            c:116092             4.0615e+08    0    0    0    0    0   
c:176998            c:163003            2.94595e+08    0    0    0    0    0   
c:150328            c:25307             2.40649e+08    0    0    0    0    0   
...                                             ...  ...  ...  ...  ...  ...   
c:83647             c:28502               266000000    0    0    0    0    0   
c:26486             c:3445                 27700000    0    0    0    0    0   
                    c:3525                 27000000    0    0    0    0    0   
c:7060              c:50055                54000000    0    0    0    0    0   
c:31279             c:31278                98300000    0    0    0    0    0   
c:1246              c:596                  45000000    0    0    0    0    0   
c:20211             c:26091              4.0615e+08    0    0    0    0    0   
c:15446             c:153534            7.05912e+08    0    0    0    0    0   
c:59312             c:228074              195000000    0    0    0    0    0   
c:31357             c:32180               175000000    0    0    0    0    0   
c:259799            c:183726               30400000    0    0    0    0    0   
c:7060              c:25775              5750000000    0    0    0    0    0   
c:276618            c:276620               64000000    0    0    0    0    0   
c:44898             c:141919               72500000    0    0    0    0    0   
c:31357             c:31356                62870000    0    0    0    0    0   
c:20211             c:35403                95000000    0    0    0    0    0   
c:4199              c:6833              2.64731e+08    0    0    0    0    0   
c:14011             c:3455             

In [160]:
df_acquiring.to_pickle('./dataframe_cleaned/df_acquiring') 

Before kicking away the negative durations 
- 9532 rows × 4 columns 

After kicking away the negative durations 
- 6866 rows × 4 columns

After dropping entries with missing values price amount
- 1875 rows × 4 columns

Current:
- 4130 rows

## Acquired

Before kicking away the negative durations 
- 9532 rows × 4 columns

After kicking away the negative durations 
- 4821 rows × 4 columns


In [165]:
df_acquired = df_acq.set_index(['acquired_object_id']).sort_index()
df_acquired.sort_values(by=['acquired_time_since_founded'],inplace=True)
df_acquired.drop(['id_x','id_y','created_at','category_code','acquisition_id','acquiring_object_id','acquired_at','acquiring_time_since_founded','term_code','price_amount','price_currency_code'],axis=1,inplace=True)
df_acquired = df_acquired[df_acquired['acquired_time_since_founded'].gt(0)]

df_acquired.to_pickle('./dataframe_cleaned/df_acquired') 
df_acquired

AUD  CAD  EUR  GBP  JPY  SEK  USD  cash  cash_and_stock  \
acquired_object_id                                                            
c:22103               0    0    0    0    0    0    1     1               0   
c:54651               0    0    0    0    0    0    1     0               0   
c:159552              0    0    0    0    0    0    1     0               0   
c:329                 0    0    0    0    0    0    1     0               0   
c:2315                0    0    0    0    0    0    1     0               0   
c:54405               0    0    0    0    0    0    1     0               0   
c:8565                0    0    0    0    0    0    1     0               0   
c:86                  0    0    0    0    0    0    1     0               0   
c:45891               0    0    0    0    0    0    1     1               0   
c:193193              0    0    0    0    0    0    1     0               0   
c:192352              0    0    0    0    0    0    1     0               0   
c:681                 0    0    0    0    0    0    1     1               0   
c:47111               0    0    0    0    0    0    1     1               0   
c:149002              0    0    0    0    0    0    1     0               0   
c:195841              0    0    0    0    0    0    1     0               0   
c:3306                0    0    0    0    0    0    1     1               0   
c:2370                0    0    0    0    0    0    1     0               1   
c:213362              0    0    0    0    0    0    1     0               0   
c:27053               0    0    0    0    0    0    1     0               0   
c:1229                0    0    0    0    0    0    1     0               1   
c:156584              0    0    0    0    0    0    1     0               0   
c:175581              0    0    0    0    0    0    1     0               0   
c:212610              0    0    0    0    0    0    1     0               0   
c:183206              0    0    0    0    0    0    1     0               0   
c:211517              0    0    0    0    0    0    1     0               0   
c:187849              0    0    0    0    0    0    1     0               0   
c:3138                0    0    0    0    0    0    1     0               0   
c:182466              0    0    0    0    0    0    1     0               0   
c:8652                0    0    0    0    0    0    1     1               0   
c:245790              0    0    0    0    0    0    1     0               0   
...                 ...  ...  ...  ...  ...  ...  ...   ...             ...   
c:30506               0    0    0    0    0    0    1     0               0   
c:269416              0    0    0    0    0    0    1     0               0   
c:38641               0    0    0    0    0    0    1     0               0   
c:242705              0    0    0    0    0    0    1     0               0   
c:216499              0    0    0    0    0    0    1     0               0   
c:25631               0    0    0    0    0    0    1     0               0   
c:32156               0    0    0    0    0    0    1     0               0   
c:29313               0    0    0    0    0    0    1     1               0   
c:267421              0    0    0    0    0    0    1     0               0   
c:5296                0    0    0    0    0    0    1     0               0   
c:34009               0    0    0    0    0    0    1     0               0   
c:2332                0    0    0    0    0    0    1     0               1   
c:51460               0    0    0    0    0    0    1     0               0   
c:81111               0    0    0    0    0    0    1     1               0   
c:86083               0    0    0    0    0    0    1     1               0   
c:10232               0    0    0    0    0    0    1     0               0   
c:270108              0    0    0    0    0    0    1     1               0   
c:220985              0    0    0    0    0    0    1     0           

## Funding


## Funding Data
- Initial: 17689 companies 31954 rows
- After: 16778 companies , 25854 rows 

In [350]:
TermList = ['angel','seed','a', 'b','c','d','e','f','g']
            # 'convertible','crowd','crowd_equity','debt_round','grant',
            # 'partial','post_ipo_debt','post_ipo_equity','private_equity','secondary_market','unattributed']


In [351]:
df_funding= pd.read_csv(data_path + 'cb_funding_rounds.csv', encoding = "ISO-8859-1")

df_funding = df_funding.replace(r'\N', np.nan)


#drop those funding rounds out of concern
df_funding = df_funding[df_funding['funding_round_code'].isin(TermList)]
df_funding = pd.concat([df_funding, pd.get_dummies(df_funding['funding_round_code'])], axis=1, sort=False)

df_funding = pd.merge(df_funding,df_object[['id','founded_at']],how='left',left_on=['object_id'],right_on=['id'])
df_funding = df_funding.drop(['source_url','created_by','pre_money_valuation_usd',
                            'pre_money_valuation','pre_money_currency_code','post_money_valuation_usd',
                             'post_money_valuation','post_money_currency_code','raised_amount','raised_currency_code'], axis=1)
df_funding

df_funding = df_funding[df_funding['founded_at'].notnull()]

df_funding['created_at'] =  pd.to_datetime(df_funding['created_at'])
df_funding.funded_at = pd.to_datetime(df_funding.funded_at)

In [352]:

def getFundDate(row):
    if row[['funded_at']].isnull().item():
        if row[['is_last_round']].item()==1 or (row[['funding_round_code']] == 'angel').item()==1:
            row[['funded_at']] = row[['founded_at']]
        elif row[['is_first_round']].item()==1 or (row[['funding_round_code']] == 'a').item()==1:
            row[['funded_at']] = row[['created_at']]
        elif (row[['funding_round_code']] =='seed').item()==1:
            pre_rows = df_funding[df_funding['object_id'] == row['object_id']]
            angel_row = pre_rows[pre_rows['funding_round_code']=='angel']
            if not angel_row.empty and angel_row['funded_at'].notnull().item()==1:
                td = row['created_at'] - angel_row['funded_at']
                td = td.astype('timedelta64[D]').values
                delta = random.randint(1,td)
                row[['funded_at']] = angel_row[['funded_at']] + pd.DateOffset(days=delta)
            else:
                row[['funded_at']] = row[['created_at']]
        elif (row[['funding_round_code']] =='b').item()==1:
            pre_rows = df_funding[df_funding['object_id'] == row['object_id']]
            angel_row = pre_rows[pre_rows['funding_round_code']=='a']
            if not angel_row.empty and angel_row['funded_at'].notnull().item()==1:
                td = row['created_at'] - angel_row['funded_at']
                td = td.astype('timedelta64[D]').values
                delta = random.randint(1,td)
                row[['funded_at']] = angel_row[['funded_at']] + pd.DateOffset(days=delta)
            else:
                row[['funded_at']] = row[['created_at']]
        elif (row[['funding_round_code']] =='c').item()==1:
            pre_rows = df_funding[df_funding['object_id'] == row['object_id']]
            angel_row = pre_rows[pre_rows['funding_round_code']=='b']
            if not angel_row.empty and angel_row['funded_at'].notnull().item()==1:
                td = row['created_at'] - angel_row['funded_at']
                td = td.astype('timedelta64[D]').values
                delta = random.randint(1,td)
                row[['funded_at']] = angel_row[['funded_at']] + pd.DateOffset(days=delta)
            else:
                row[['funded_at']] = row[['created_at']]
    return row

df_funding = df_funding.apply(getFundDate,axis=1)
                
                
                
                
            

In [353]:
## make timestamp 

df_funding['time_since_founded'] = pd.Timedelta('nan')

def getDuration(row):
    
    #df_row = df_object.loc[df_object['id'] == row.object_id]
    #if not df_row.empty:
    if pd.notnull(row['founded_at']) and pd.notnull(row['funded_at']):
    #        duration = row.funded_at - df_row.founded_at
        duration = row.funded_at - row.founded_at
        row['time_since_founded'] = pd.to_timedelta([duration]).astype('timedelta64[D]').values
    return row

df_funding = df_funding.apply(getDuration,axis=1)

#only positive date
df_funding = df_funding[df_funding['time_since_founded'].gt(-0.1)]
df_funding['time_since_founded'] = df_funding['time_since_founded'].astype(int)

In [354]:
df_funding.sort_values(by=['time_since_founded'],inplace=True)
df_funding

id_x  funding_round_id object_id  funded_at funding_round_type  \
16944  26956             26956  c:144040 2008-03-01              angel   
4093    4962              4962   c:18038 2006-09-01           series-a   
19165  30642             30642  c:161452 2012-01-01              angel   
19166  30643             30643  c:161453 2012-01-01              angel   
19169  30648             30648  c:161480 2012-02-01              angel   
28530  48635             48635  c:264245 2009-07-01              angel   
4121    5004              5004   c:18202 2008-04-01              angel   
19187  30677             30677   c:51578 2009-01-01              angel   
19191  30683             30683  c:161699 2011-06-01              angel   
19212  30717             30717  c:161944 2012-03-01              angel   
4145    5036              5036    c:4299 2003-01-01              angel   
28502  48542             48542  c:263989 2011-01-01              angel   
13601  21419             21419   c:68549 2011-02-11              angel   
13592  21396             21396   c:68486 2009-01-01              angel   
13587  21387             21387   c:68442 2010-01-01              angel   
19232  30750             30750  c:162067 2005-01-01              angel   
4184    5083              5083   c:18701 2008-10-01              angel   
4188    5088              5088   c:18741 2008-05-01              angel   
28549  48672             48672  c:264553 2012-10-01              angel   
13574  21361             21361   c:68329 2010-01-01              angel   
19162  30639             30639   c:45196 2009-09-01              angel   
13671  21536             21536   c:69063 2009-09-01              angel   
28705  49127             49127  c:265497 2012-10-11              angel   
19125  30579             30579  c:161240 2010-10-01           series-a   
28693  49092             49092  c:265463 2013-06-01              angel   
13719  21622             21622   c:64996 2010-01-01              angel   
4009    4848              4848   c:17758 2006-09-06           series-a   
28677  49041             49041  c:265370 2013-06-01              angel   
13714  21614             21614   c:69596 1998-01-01              angel   
28661  48995             48995  c:265241 2013-03-01              angel   
...      ...               ...       ...        ...                ...   
29328  50646             50646  c:145945 2013-10-02              angel   
12466  19369             19369   c:59556 2010-10-20           series-a   
24552  39184             39184  c:220572 2013-03-28              angel   
26876  44380             44380  c:243852 2012-07-30              angel   
24374  38683             38683   c:69136 2013-01-01              angel   
24048  37994             37994  c:212776 2013-05-17           series-a   
26981  44624             44624  c:244910 2013-08-02              angel   
26656  43840             43840   c:16243 2012-08-31              angel   
30599  53862             53862  c:277196 2013-11-05              angel   
2369    2695              2695    c:6713 2008-02-05           series-b   
27021  44714             44714  c:245323 2012-07-13           series-a   
7110    9542              9542   c:35691 2008-02-02           series-a   
24472  38927             38927  c:219131 2013-04-12              angel   
29135  50207             50207  c:267819 2011-09-14           series-a   
25550  41332             41332  c:177075 2012-06-13              angel   
24530  39100             39100  c:177075 2013-05-30              angel   
17041  27122             27122  c:144811 2011-12-16           series-a   
11723  18188             18188   c:36848 2010-09-09          series-c+   
24769  39615             39615  c:164565 2013-06-07              angel   
30822  54518             54518  c:279000 2013-11-12           series-a   
12601  19634             19634    c:8261 2005-12-30          series-c+   
25889  42182             42182  c:236502 2013-07-09              angel   
29459  50924

## Average fund each term

In [355]:

avg_fund = []
for term in TermList:
    row = df_funding.loc[df_funding['funding_round_code']==term,['raised_amount_usd']]
    fund = row[pd.notnull(row.raised_amount_usd)].astype('float128').mean()
    avg_fund.append(fund)
fund_dic = dict(zip(TermList,avg_fund))

def getFund(row):
    if row[['raised_amount_usd']].isnull().item() and row[['funding_round_code']].notnull().item():
        row[['raised_amount_usd']] = fund_dic[row['funding_round_code']]
    return row
df_funding = df_funding.apply(getFund,axis=1)

In [356]:

df_funding.drop(['id_x','id_y','created_at','updated_at','funding_round_id','funding_round_type','founded_at','funded_at','is_first_round','is_last_round'],axis=1,inplace=True)


In [359]:
df_funding.to_pickle('./dataframe_cleaned/df_funding')

In [331]:
len(df_funding.groupby('object_id').groups)

16778

In [360]:
null_columns = df_funding.columns[df_funding.isnull().any()]
print(df_funding[null_columns].isnull().sum())
df_funding

Series([], dtype: float64)


object_id funding_round_code raised_amount_usd  participants  a  angel  \
16944  c:144040               seed           1000000             0  0      0   
4093    c:18038                  a          13350000             4  1      0   
19165  c:161452               seed            866033             0  0      0   
19166  c:161453               seed            100000             0  0      0   
19169  c:161480              angel            250000             1  0      1   
28530  c:264245              angel            832099             1  0      1   
4121    c:18202               seed            425000             2  0      0   
19187   c:51578              angel            695850             0  0      1   
19191  c:161699               seed            575000             1  0      0   
19212  c:161944              angel           1000000             5  0      1   
4145     c:4299               seed           1000000             0  0      0   
28502  c:263989               seed            650000             0  0      0   
13601   c:68549               seed             50000             0  0      0   
13592   c:68486              angel            832099             0  0      1   
13587   c:68442               seed            866033             1  0      0   
19232  c:162067               seed              1000             0  0      0   
4184    c:18701               seed            100000             0  0      0   
4188    c:18741              angel            279720             6  0      1   
28549  c:264553               seed            250000             0  0      0   
13574   c:68329              angel            200000             0  0      1   
19162   c:45196               seed            143140             1  0      0   
13671   c:69063               seed            866033             0  0      0   
28705  c:265497               seed            866033             0  0      0   
19125  c:161240               seed           1500000             0  0      0   
28693  c:265463               seed             50000             0  0      0   
13719   c:64996               seed           1150000             1  0      0   
4009    c:17758                  a           4000000             1  1      0   
28677  c:265370               seed             50000             0  0      0   
13714   c:69596               seed            200000             0  0      0   
28661  c:265241               seed            866033             1  0      0   
...         ...                ...               ...           ... ..    ...   
29328  c:145945               seed           1000000             0  0      0   
12466   c:59556                  a       6.85208e+06             1  1      0   
24552  c:220572               seed           1158000             0  0      0   
26876  c:243852               seed            500000             0  0      0   
24374   c:69136               seed            866033             0  0      0   
24048  c:212776                  a          12000000             1  1      0   
26981  c:244910               seed            725000             0  0      0   
26656   c:16243               seed            127890             0  0      0   
30599  c:277196               seed             86000             0  0      0   
2369     c:6713                  b          11750000             3  0      0   
27021  c:245323                  a             50000             0  1      0   
7110    c:35691                  a           6000000             0  1      0   
24472  c:219131               seed             43120             0  0      0   
29135  c:267819               seed           1800000             0  0      0   
25550  c:177075               seed           1000000             0  0      0   
24530  c:177075               seed            675000             0  0      0   
17041  c:144811                  a          10000000             1  1      0   
11723   c:36848                  d          65000000             3  0      0   
24769 